gunrock/graph/dynamic_graph/dynamic_graph_parallel_iterator.cuh

```diff
@@ -0,0 +1,162 @@
+// ----------------------------------------------------------------------------
+// Gunrock -- Fast and Efficient GPU Graph Library
+// ----------------------------------------------------------------------------
+// This source code is distributed under the terms of LICENSE.TXT
+// in the root directory of this source distribution.
+// ----------------------------------------------------------------------------
+
+/**
+ * @file
+ * dynamic_graph_parallel_iterator.cuh
+ *
+ * @brief Dynamic Graph Data Structure iterator
+ */
+#pragma once
+
+#include <gunrock/util/array_utils.cuh>
+#include <gunrock/graph/csr.cuh>
+#include <gunrock/graph/graph_base.cuh>
+
+#include <gunrock/graph/dynamic_graph/dynamic_graph_base.cuh>
+#include <gunrock/graph/dynamic_graph/dynamic_graph_unweighted.cuh>
+#include <gunrock/graph/dynamic_graph/dynamic_graph_weighted.cuh>
+
+namespace gunrock {
+namespace graph {
+
+template <typename VertexT, typename ValueT, bool IsMap>
+struct nodeReader {
+  __device__ static void readNode(const uint32_t& key, VertexT& eId,
+                                  ValueT& eVal, const int& laneId) {
+    eVal = __shfl_xor_sync(0xFFFFFFFF, key, 1);
+    bool key_lane = !(laneId % 2) && (laneId < 30);
+    eId = key_lane ? key : EMPTY_KEY;
+    eVal = key_lane ? eVal : EMPTY_KEY;
+    eId = (key != EMPTY_KEY) ? eId
+                             : util::PreDefinedValues<VertexT>::InvalidValue;
+    eVal = (key != EMPTY_KEY) ? eVal
+                              : util::PreDefinedValues<ValueT>::InvalidValue;
+  }
+};
+template <typename VertexT, typename ValueT>
+struct nodeReader<VertexT, ValueT, false> {
+  __device__ static void readNode(const uint32_t& key, VertexT& eId,
+                                  ValueT& eVal, const int& laneId) {
+    bool key_lane = laneId < 30;
+    eId = key_lane ? key : EMPTY_KEY;
+    bool is_valid = (key != EMPTY_KEY);
+    eId = (key != EMPTY_KEY) ? eId
+                             : util::PreDefinedValues<VertexT>::InvalidValue;
+    eVal = util::PreDefinedValues<ValueT>::InvalidValue;
+  }
+};
```

Comment:
Are other users expected to use these or are they for low level internal use only? Should they be part of a private interface?

Reply:
Yeah, you are right. I should move all the iterator code to `SlabHashGraphMap/Set/Base` files and this file will only contain calls to these functions.